domain/application/service/storage.go

```diff
@@ -334,6 +391,166 @@ func (s *Service) DetachStorage(ctx context.Context, storageID corestorage.ID) e
 }
 
+// getUnitStorageInfo retrieves the existing storage information for a unit.
+// This includes the storage directives set for the unit and also any storage
+// that is currently owned by the unit.
+//
+// The following errors can be expected:
+// - [applicationerrors.UnitNotFound] when the unit no longer exists.
+func (s *ProviderService) getUnitStorageInfo(
+	ctx context.Context,
+	unitUUID coreunit.UUID,
+) (
+	[]application.StorageDirective,
+	map[domainstorage.Name][]domainstorage.StorageInstanceUUID,
+	error,
+) {
+	existingUnitStorage, err := s.st.GetUnitOwnedStorageInstances(
+		ctx, unitUUID,
+	)
+	if err != nil {
+		return nil, nil, errors.Errorf(
+			"getting unit %q owned storage instances: %w",
+			unitUUID, err,
+		)
+	}
+
+	storageDirectives, err := s.st.GetUnitStorageDirectives(ctx, unitUUID)
+	if err != nil {
+		return nil, nil, errors.Errorf(
+			"getting unit %q storage directives: %w",
+			unitUUID, err,
+		)
+	}
+
+	return storageDirectives, existingUnitStorage, nil
+}
+
+// makeUnitStorageArgs creates the storage arguments required for a CAAS unit in
+// the model. This func looks at the set of directives for the unit and the
+// existing storage available. From this any new instances that need to be
+// created are calculated and all storage attachments are added.
+func makeUnitStorageArgs(
+	storageDirectives []application.StorageDirective,
+	existingStorage map[domainstorage.Name][]domainstorage.StorageInstanceUUID,
+) (application.CreateUnitStorageArg, error) {
+	rvalDirectives := make([]application.CreateUnitStorageDirectiveArg, 0, len(storageDirectives))
+	rvalInstances := []application.CreateUnitStorageInstanceArg{}
+	rvalToAttach := make([]domainstorage.StorageInstanceUUID, 0, len(storageDirectives))
+	rvalToOwn := make([]domainstorage.StorageInstanceUUID, 0, len(storageDirectives))
+
+	for _, sd := range storageDirectives {
+		// We make the storage directive args first. This is becase we know we
+		// will change the count in the struct later.
+		rvalDirectives = append(rvalDirectives, application.CreateUnitStorageDirectiveArg{
+			Count:        sd.Count,
+			Name:         sd.Name,
+			PoolUUID:     sd.PoolUUID,
+			ProviderType: sd.ProviderType,
+			Size:         sd.Size,
+		})
+
+		existingStorageUUIDs := existingStorage[sd.Name]
+		numExistingStorage := uint32(len(existingStorageUUIDs))
+		if numExistingStorage > sd.Count {
+			// A storage directive only supports n number of storage instances
+			// per directive. If there exists more existing storage for this
+			// directive than supported, we return an error.
+			//
+			// This is undefined behaviour in the Juju modeling.
+			return application.CreateUnitStorageArg{}, errors.Errorf(
+				"unable to use %d existing storage instances for directive %q, greater than supported count %d",
+				numExistingStorage, sd.Name, sd.Count,
+			)
+		}
+		// Remove the already existing storage instances from the count.
+		sd.Count -= numExistingStorage
```

Comment:
This is scary. This will wrap around if `numExistingStorage` is too large creating very serious issues. I think the if is safe, but it's still scary mary.

Reply:
Oh I agree 100%. I moved this here, to get it close as possible to the if statement to protect it from underflow and make it obvious to the reader. I've now put in a test before casting the length to a uint32.

IMO, we should never not use an `int`. A signed, 64bit integer is only half as small as an unsigned 64 bit integer, but behaves how people expect, because you are never dancing around one of the limits for most needs.